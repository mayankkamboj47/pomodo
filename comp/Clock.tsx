import { AppState, dispatchType } from "@/utils/types";

interface propTypes {
  time: AppState["time"];
  type: AppState["clockType"];
  status: AppState["clockStatus"];
  totalMins: AppState["totalMins"];
  dispatch: dispatchType;
}

const Clock = ({ time, type, status, totalMins, dispatch }: propTypes) => {
  const onPausePlay = (e: React.MouseEvent<HTMLButtonElement>) => {
    dispatch({ type: "clock.stop-resume" });
    e.stopPropagation();
  };
  return (
    <div className="clock absolute" data-testid="clock">
      <div className="w-40 h-40 shadow-2xl rounded-full bg-blue-50"></div>
      <div className="absolute top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 flex-column justify-center align-center text-center w-40">
        <div className="m-0 p-0">{type}</div>
        <TimeDisplay time={time} totalMins={totalMins} dispatch={dispatch} />
        <button className="" onClick={onPausePlay}>
          {status === "running" ? (
            <>
              {pauseSvg}
              <span className="sr-only">Pause</span>
            </>
          ) : (
            <>
              {playSvg}
              <span className="sr-only">Play</span>
            </>
          )}
        </button>
      </div>
    </div>
  );
};

function TimeDisplay({
  time,
  totalMins,
  dispatch,
}: {
  time: number;
  totalMins: number;
  dispatch: dispatchType;
}) {
  return time === totalMins * 60 ? (
    <div className="">
      <label className="sr-only" htmlFor="clockTime">Time in minutes</label>
      <input
        type="number"
        name="clockTime"
        className="w-11 text-2xl bg-blue-50"
        value={totalMins}
        min={1}
        onChange={(e) =>
          dispatch({ type: "clock.setTime", time: Number(e.target.value) })
        }
      />
      minutes
    </div>
  ) : (
    <div className="text-2xl font-bold">{clocktimeToString(time)}</div>
  );
}
function clocktimeToString(clockTime: number) {
  let minutes = Math.floor(clockTime / 60);
  let seconds = clockTime % 60;
  return `${minutes < 10 ? "0" : ""}${minutes}:${
    seconds < 10 ? "0" : ""
  }${seconds}`;
}
export default Clock;

// helper variables

const pauseSvg = (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="32"
    height="32"
    fill="black"
    viewBox="0 0 16 16"
  >
    <path d="M6 3.5a.5.5 0 0 1 .5.5v8a.5.5 0 0 1-1 0V4a.5.5 0 0 1 .5-.5zm4 0a.5.5 0 0 1 .5.5v8a.5.5 0 0 1-1 0V4a.5.5 0 0 1 .5-.5z" />
  </svg>
);

const playSvg = (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="32"
    height="32"
    fill="black"
    viewBox="0 0 16 16"
  >
    <path d="M10.804 8 5 4.633v6.734L10.804 8zm.792-.696a.802.802 0 0 1 0 1.392l-6.363 3.692C4.713 12.69 4 12.345 4 11.692V4.308c0-.653.713-.998 1.233-.696l6.363 3.692z" />
  </svg>
);
