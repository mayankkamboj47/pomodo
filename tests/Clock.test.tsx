import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Clock from '../comp/Clock'

// Mock dispatch function
const dispatch = jest.fn() 

describe('Clock', () => {

  beforeEach(() => {
    dispatch.mockClear()
  })

  it('renders correctly', () => {
    const { asFragment } = render(<Clock endTime={100} dispatch={dispatch} />)
    expect(asFragment()).toMatchSnapshot()
  })

  describe('display', () => {

    it('renders time remaining', () => {
      render(<Clock endTime={100} dispatch={dispatch}/>)
      expect(screen.getByText('1:40')).toBeInTheDocument()
    })

    it('renders time in mm:ss format', () => {
      render(<Clock endTime={61} dispatch={dispatch}/>)
      expect(screen.getByText('1:01')).toBeInTheDocument()
    })

  })

  describe('countdown', () => {

    it('counts down remaining time', async () => {
      jest.useFakeTimers()
      render(<Clock endTime={100} status="running" dispatch={dispatch}/>)

      jest.advanceTimersByTime(2000)
      expect(screen.getByText('1:38')).toBeInTheDocument()
    })

    it('pauses countdown when status is not running', async () => {
      jest.useFakeTimers()
      render(<Clock endTime={100} status="paused" dispatch={dispatch}/>)

      jest.advanceTimersByTime(2000)
      expect(screen.getByText('1:40')).toBeInTheDocument() 
    })

    it('clears interval on unmount', () => {
      jest.useFakeTimers()
      const { unmount } = render(<Clock endTime={100} status="running" dispatch={dispatch}/>)
      unmount()

      expect(clearInterval).toHaveBeenCalledTimes(1)
    })

  })

  describe('completion', () => {

    it('dispatches beep when countdown completes', async () => {
      render(<Clock endTime={0} status="running" dispatch={dispatch}/>)

      jest.advanceTimersByTime(1000)
      expect(dispatch).toHaveBeenCalledWith({type: 'beep'})
    })

    it('clears interval when countdown completes', async () => {
      jest.useFakeTimers()
      render(<Clock endTime={0} status="running" dispatch={dispatch}/>)

      jest.advanceTimersByTime(1000)
      expect(clearInterval).toHaveBeenCalledTimes(1)
    })

  })

  describe('interactions', () => {

    it('dispatches stop-resume on button click', () => {
      render(<Clock endTime={100} status="running" dispatch={dispatch}/>)

      userEvent.click(screen.getByRole('button'))
      expect(dispatch).toHaveBeenCalledWith({
        type: 'clock.stop-resume', 
        clockTime: 100  
      })
    })

    it('toggles pause icon on button click', async () => {
      render(<Clock endTime={100} status="running" dispatch={dispatch}/>)

      expect(screen.getByTestId('play-icon')).toBeInTheDocument()

      userEvent.click(screen.getByRole('button'))

      expect(screen.getByTestId('pause-icon')).toBeInTheDocument()
    })

  })

})