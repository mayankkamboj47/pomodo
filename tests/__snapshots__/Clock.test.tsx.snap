// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clock renders correctly 1`] = `
<DocumentFragment>
  <div
    class="absolute"
  >
    <div
      class="w-40 h-40 shadow-2xl rounded-full bg-blue-50"
    />
    <div
      class="absolute top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 flex-column justify-center align-center text-center w-40"
    >
      <div
        class="m-0 p-0"
      />
      <div
        class="text-2xl font-bold"
      >
        -28236300:0-44
      </div>
      <button
        class=""
      >
        <svg
          fill="black"
          height="32"
          viewBox="0 0 16 16"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.804 8 5 4.633v6.734L10.804 8zm.792-.696a.802.802 0 0 1 0 1.392l-6.363 3.692C4.713 12.69 4 12.345 4 11.692V4.308c0-.653.713-.998 1.233-.696l6.363 3.692z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
